sled-agent/src/common/instance.rs

```diff
@@ -288,6 +224,55 @@ impl InstanceStates {
         &mut self,
         observed: &ObservedPropolisState,
     ) -> Option<Action> {
+        fn transition_migration(
+            current: &mut Option<MigrationRuntimeState>,
+            ObservedMigrationState { id, state }: ObservedMigrationState,
+            now: DateTime<Utc>,
+        ) {
+            if let Some(ref mut m) = current {
+                // Don't generate spurious state updates if the migration is already in
+                // the state we're transitioning to.
+                if m.migration_id == id && m.state == state {
+                    return;
+                }
+                m.state = state;
+                if m.migration_id == id {
+                    m.gen = m.gen.next();
+                } else {
+                    m.migration_id = id;
+                    m.gen = Generation::new().next();
+                }
```

Comment:
I'm a little surprised to see this, but trying to grok it fully:

If this was checking whether or not `m.migration_id` was `None`, and it set `m.migration_id = id` in that case, I'd get it, but this appears to be changing a non-None ID to a different value. When could that happen? Would that be, like, switching from "one migration attempt" to "another migration attempt"?
(If that's the case, does anything stop old migration requests from overwriting newer migration attempts? The action of "resetting the generation number" here makes me nervous)

Reply:
We would, potentially, do this if we had attempted to migrate out of this VMM, and that migration had failed (e.g. due to the target of the migration dying unexpectedly). In that case, we might then start a *new* migration to a new target VMM, since we would still want to get the instance out of this one. In that case, we'll continue to report the most recent migration state (the previous, failed migration) until a new one starts, so we don't clear it out immediately, because we want to actually make sure it's been observed by nexus. Thus, we just clobber it here.

Reply:
Gotcha. I was debating if this would be a problem in a scenario with "delayed network messages" -- e.g., it would be wrong to clobber the value of `current` with an old observation -- but I don't *think* that's possible with how `apply_propolis_observation` is being invoked (it happens in response to monitoring propolis, but the observations should be ordered, since this is all happening within the sled).